tempesta_fw/sock.c

```diff
@@ -866,17 +866,21 @@ ss_tcp_data_ready(struct sock *sk)
 		if (ss_tcp_process_data(sk) &&
 		    !(SS_CONN_TYPE(sk) & Conn_Stop)) {
 			/*
-			 * Drop connection in case of internal errors,
+			 * Close connection in case of internal errors,
 			 * banned packets, or FIN in the received packet,
 			 * and only if it's not on hold until explicitly
 			 * closed.
 			 *
-			 * ss_linkerror() is responsible for calling
+			 * ss_close() is responsible for calling
 			 * application layer connection closing callback.
 			 * The callback will free all SKBs linked with
 			 * the message that is currently being processed.
+			 *
+			 * Use ss_close() to close the connection after
+			 * sending responses in the queue, for example,
+			 * tls alerts
 			 */
-			ss_linkerror(sk);
+			ss_close(sk, SS_F_SYNC);
```

Comment:
I reviewed the code around once more and it seems that we should do more complex changes. `ss_tcp_process_data() -> `ss_tcp_process_skb()` returns true/false on all kind of problems: e.g. inability to unroll an skb and a security event, like password cracking. However, the problems are very different - we should send pending data in first case and send RST (subject for #861) in the second. However, it's proposed to extend `ss_close()` with an argument for RST and this change just makes step further. In #861 we'll need to amend `ss_tcp_process_skb()` to return SS_BAD in generic errors and SS_DROP on security events. So I'm still fine with the change and will just add the requirement to #861.

Please add TODO #861 with the discussion summary to the comment and put coma at the end of comment.